core/network/src/protocol/sync.rs

```diff
@@ -441,6 +441,53 @@ impl<B: BlockT> ChainSync<B> {
 		})
 	}
 
+	/// Request syncing for the given block from given set of peers.
+	// This is similar to on_block_announce with unknown parent hash.
+	pub fn sync_fork(&mut self, peers: Vec<PeerId>, hash: &B::Hash, number: NumberFor<B>) ->
+		Vec<(PeerId, BlockRequest<B>)>
+	{
+		debug!(target: "sync", "Explicit sync request for block {:?} with {:?}", hash, peers);
+		if self.is_known(&hash) || self.is_already_downloading(&hash) {
+			debug!(target: "sync", "Refusing to sync known hash {:?}", hash);
+			return Vec::default();
+		}
+
+		let block_status = self.client.block_status(&BlockId::Number(number - One::one()))
+			.unwrap_or(BlockStatus::Unknown);
+		if block_status == BlockStatus::InChainPruned {
+			trace!(target: "sync", "Refusing to sync ancient block {:?}", hash);
+			return Vec::default();
+		}
+
+		self.is_idle = false;
+		let mut requests = Vec::new();
+		for peer_id in peers {
+			let peer = if let Some(peer) = self.peers.get_mut(&peer_id) {
+				peer
+			} else {
+				debug!(target: "sync", "Called sync_fork with a bad peer ID");
+				continue;
+			};
+			if let PeerSyncState::AncestorSearch(_, _) = peer.state {
+				continue;
+			}
+
+			if number > peer.best_number {
+				peer.best_number = number;
+				peer.best_hash = hash.clone();
```

Comment:
why set that to the peer's best block?

Reply:
`best_number` is used to track global sync target. `best_hash` is not used for anything really, other than reporting generic peer info.